File: Sources/CodyUsageCore/RateLimitParser.swift
import Foundation

public enum RateLimitParserError: Error { case invalidPayload }

public enum RateLimitParser {
    public static func parse(data: Data) throws -> RateLimitResult {
        try parse(JSONSerialization.jsonObject(with: data))
    }

    public static func parse(_ value: Any) throws -> RateLimitResult {
        guard let root = value as? [String: Any] else { throw RateLimitParserError.invalidPayload }
        var snapshots: [[String: Any]] = []
        if let buckets = root["rateLimitsByLimitId"] as? [String: Any],
           let codex = buckets["codex"] as? [String: Any] {
            snapshots.append(codex)
        } else if let single = root["rateLimits"] as? [String: Any] {
            snapshots.append(single)
        } else if let buckets = root["rateLimitsByLimitId"] as? [String: Any] {
            snapshots.append(contentsOf: buckets.values.compactMap { $0 as? [String: Any] })
        }

        var windows: [RateLimitWindow] = []
        for snapshot in snapshots {
            for key in ["primary", "secondary"] {
                guard let raw = snapshot[key] as? [String: Any],
                      let used = integer(raw["usedPercent"] ?? raw["used_percent"]) else { continue }
                let duration = integer(raw["windowDurationMins"] ?? raw["window_minutes"])
                let resetSeconds = integer(raw["resetsAt"] ?? raw["resets_at"])
                windows.append(RateLimitWindow(
                    usedPercent: used,
                    windowDurationMinutes: duration,
                    resetsAt: resetSeconds.map { Date(timeIntervalSince1970: TimeInterval($0)) }
                ))
            }
        }

        let unique = Array(Set(windows.map(WindowKey.init))).map(\.window)
        return RateLimitResult(
            fiveHour: closest(in: unique, target: 300, tolerance: 180),
            weekly: closest(in: unique, target: 10_080, tolerance: 2_880)
        )
    }

    private static func integer(_ value: Any?) -> Int? {
        if let int = value as? Int { return int }
        if let number = value as? NSNumber { return number.intValue }
        if let string = value as? String { return Int(string) }
        return nil
    }

    private static func closest(in windows: [RateLimitWindow], target: Int, tolerance: Int) -> RateLimitWindow? {
        windows
            .filter { $0.windowDurationMinutes != nil && abs($0.windowDurationMinutes! - target) <= tolerance }
            .min { abs($0.windowDurationMinutes! - target) < abs($1.windowDurationMinutes! - target) }
    }

    private struct WindowKey: Hashable {
        let used: Int
        let duration: Int?
        let reset: Int?
        let window: RateLimitWindow
        init(_ window: RateLimitWindow) {
            used = window.usedPercent
            duration = window.windowDurationMinutes
            reset = window.resetsAt.map { Int($0.timeIntervalSince1970) }
            self.window = window
        }
        static func == (lhs: WindowKey, rhs: WindowKey) -> Bool {
            lhs.used == rhs.used && lhs.duration == rhs.duration && lhs.reset == rhs.reset
        }
        func hash(into hasher: inout Hasher) {
            hasher.combine(used); hasher.combine(duration); hasher.combine(reset)
        }
    }
}

File: windows/src/CodyUsageOverlay.Core/RateLimitParser.cs
using System.Text.Json;

namespace CodyUsageOverlay.Core;

public static class RateLimitParser
{
    public static RateLimitResult Parse(ReadOnlySpan<byte> json)
    {
        using var document = JsonDocument.Parse(json);
        return Parse(document.RootElement);
    }

    public static RateLimitResult Parse(JsonElement root)
    {
        var snapshots = new List<JsonElement>();
        if (root.TryGetProperty("rateLimitsByLimitId", out var buckets) && buckets.ValueKind == JsonValueKind.Object)
        {
            if (buckets.TryGetProperty("codex", out var codex) && codex.ValueKind == JsonValueKind.Object) snapshots.Add(codex);
            else snapshots.AddRange(buckets.EnumerateObject().Where(x => x.Value.ValueKind == JsonValueKind.Object).Select(x => x.Value));
        }
        else if (root.TryGetProperty("rateLimits", out var single) && single.ValueKind == JsonValueKind.Object) snapshots.Add(single);
        else if (root.TryGetProperty("primary", out _)) snapshots.Add(root);

        var windows = new List<RateLimitWindow>();
        foreach (var snapshot in snapshots)
        foreach (var key in new[] { "primary", "secondary" })
        {
            if (!snapshot.TryGetProperty(key, out var raw) || raw.ValueKind != JsonValueKind.Object) continue;
            var used = Integer(raw, "usedPercent", "used_percent");
            if (used is null) continue;
            var duration = Integer(raw, "windowDurationMins", "window_minutes");
            var reset = Integer64(raw, "resetsAt", "resets_at");
            windows.Add(new RateLimitWindow(used.Value, duration, reset is null ? null : DateTimeOffset.FromUnixTimeSeconds(reset.Value)));
        }

        RateLimitWindow? closest(int minutes, int tolerance) => windows
            .Where(x => x.WindowDurationMinutes is not null && Math.Abs(x.WindowDurationMinutes.Value - minutes) <= tolerance)
            .OrderBy(x => Math.Abs(x.WindowDurationMinutes!.Value - minutes)).FirstOrDefault();
        return new RateLimitResult(closest(300, 60), closest(10_080, 1440));
    }

    private static int? Integer(JsonElement element, params string[] names)
    {
        var value = Integer64(element, names);
        return value is >= int.MinValue and <= int.MaxValue ? (int)value.Value : null;
    }

    private static long? Integer64(JsonElement element, params string[] names)
    {
        foreach (var name in names)
            if (element.TryGetProperty(name, out var property))
            {
                if (property.TryGetInt64(out var number)) return number;
                if (property.ValueKind == JsonValueKind.String && long.TryParse(property.GetString(), out number)) return number;
            }
        return null;
    }
}
